src/mito2/src/gc.rs

```diff
@@ -22,41 +22,82 @@
 use object_store::{Entry, Lister};
 use serde::{Deserialize, Serialize};
 use snafu::{OptionExt, ResultExt as _, ensure};
-use store_api::storage::{FileId, RegionId};
+use store_api::storage::{FileId, FileRefsManifest, GcReport, RegionId};
+use tokio::sync::{OwnedSemaphorePermit, TryAcquireError};
 use tokio_stream::StreamExt;
 
 use crate::access_layer::AccessLayerRef;
 use crate::cache::CacheManagerRef;
 use crate::config::MitoConfig;
 use crate::error::{
     DurationOutOfRangeSnafu, EmptyRegionDirSnafu, JoinSnafu, OpenDalSnafu, RegionNotFoundSnafu,
-    Result, UnexpectedSnafu,
+    Result, TooManyGcJobsSnafu, UnexpectedSnafu,
 };
 use crate::manifest::manager::{RegionManifestManager, RegionManifestOptions, RemoveFileOptions};
 use crate::manifest::storage::manifest_compress_type;
-use crate::metrics::GC_FILE_CNT;
+use crate::metrics::GC_DEL_FILE_CNT;
 use crate::region::opener::new_manifest_dir;
 use crate::sst::file::delete_files;
-use crate::sst::file_ref::TableFileRefsManifest;
 use crate::sst::location::{self, region_dir_from_table_dir};
 
-#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
-pub struct GcReport {
-    /// deleted files per region
-    pub deleted_files: HashMap<RegionId, Vec<FileId>>,
-    /// Regions that need retry in next gc round, usually because their tmp ref files are outdated
-    pub need_retry_regions: HashSet<RegionId>,
+/// Limit the amount of concurrent GC jobs on the datanode
+pub struct GcLimiter {
+    pub gc_job_limit: Arc<tokio::sync::Semaphore>,
+    gc_concurrency: usize,
 }
 
-#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
-pub struct FileGcOption {
+pub type GcLimiterRef = Arc<GcLimiter>;
+
+impl GcLimiter {
+    pub fn new(gc_concurrency: usize) -> Self {
+        Self {
+            gc_job_limit: Arc::new(tokio::sync::Semaphore::new(gc_concurrency)),
+            gc_concurrency,
+        }
+    }
+
+    pub fn running_gc_tasks(&self) -> u32 {
+        (self.gc_concurrency - self.gc_job_limit.available_permits()) as u32
+    }
+
+    pub fn gc_concurrency(&self) -> u32 {
+        self.gc_concurrency as u32
+    }
+
+    pub fn gc_stat(&self) -> GcStat {
+        GcStat::new(self.running_gc_tasks(), self.gc_concurrency())
+    }
+
+    /// Try to acquire a permit for a GC job.
+    ///
+    /// If no permit is available, returns an `TooManyGcJobs` error.
+    pub fn permit(&self) -> Result<OwnedSemaphorePermit> {
+        self.gc_job_limit
+            .clone()
+            .try_acquire_owned()
+            .map_err(|e| match e {
+                TryAcquireError::Closed => UnexpectedSnafu {
+                    reason: format!("Failed to acquire gc permit: {e}"),
+                }
+                .build(),
+                TryAcquireError::NoPermits => TooManyGcJobsSnafu {}.build(),
+            })
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+pub struct GcConfig {
+    /// Whether GC is enabled.
+    pub enabled: bool,
```

Comment:
nitpick: Maybe `enable` is better because `enabled` is the current status. We use enable_xxx in other options.
```suggestion
    pub enable: bool,
```